bentoml/_internal/configuration/containers.py

```diff
@@ -223,24 +224,21 @@ def tracer(
 
             return NoopTracer()
 
-    logging_file_directory = t.cast(
-        Provider[t.Callable[[], str]],
+    logging_file_directory = providers.Factory[str](
+        lambda default, customized: customized or default,
```

Comment:
we should decide whether we're ok with using `or` like this.

Reply:
I prefer specifically checking `is None`

Reply:
Actually I don't know why the default value get so twisted. But it's not about the current PR.